core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -434,21 +435,48 @@ class TransactionCoordinator(brokerId: Int,
 
   private def abortTimedOutTransactions(): Unit = {
     txnManager.timedOutTransactions().foreach { txnIdAndPidEpoch =>
-      handleEndTransaction(txnIdAndPidEpoch.transactionalId,
-        txnIdAndPidEpoch.producerId,
-        txnIdAndPidEpoch.producerEpoch,
-        TransactionResult.ABORT,
-        (error: Errors) => error match {
-          case Errors.NONE =>
-            info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
-          case Errors.INVALID_PRODUCER_ID_MAPPING |
-               Errors.INVALID_PRODUCER_EPOCH |
-               Errors.CONCURRENT_TRANSACTIONS =>
-            debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
-          case e =>
-            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
-        })
-    }
+      val coordinatorEpochAndMetadata = txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
+        case None =>
+          fatal(s"Could not find transaction metadata when trying to timeout transaction with transactionalId " +
+            s"${txnIdAndPidEpoch.transactionalId}. ProducerId: ${txnIdAndPidEpoch.producerId}. ProducerEpoch: " +
+            s"${txnIdAndPidEpoch.producerEpoch}")
+          Left(Errors.INVALID_TXN_STATE)
+
+        case Some(epochAndTxnMetadata) =>
+          if (epochAndTxnMetadata.transactionMetadata.producerId != txnIdAndPidEpoch.producerId) {
+            fatal(s"Found incorrect producerId when expiring transactionalId: ${txnIdAndPidEpoch.transactionalId}. " +
+              s"Expected producerId: ${txnIdAndPidEpoch.producerId}. Found producerId: " +
+              s"${epochAndTxnMetadata.transactionMetadata.producerId}")
+            Left(Errors.INVALID_PRODUCER_ID_MAPPING)
+          } else {
+            Right(epochAndTxnMetadata)
+          }
+     }
+
+      coordinatorEpochAndMetadata match {
+        case Left(error) =>
+          logger.error(s"Could not retrieve transaction metadata for transactionalId ${txnIdAndPidEpoch.transactionalId} due " +
+            s"to error: $error")
+        case Right(epochAndMetadata) =>
+          val txnMetadata = epochAndMetadata.transactionMetadata
+          txnMetadata.inLock(txnMetadata.fenceProducerEpoch())
```

Comment:
I have emulated the locking strategy in `TransactionCoordinator.handleInitProducerId` where the fencing and the update to the log happen with different lock acquisitions. However, I am not convinced this is correct. If it isn't we may want to fix that in a future PR.

Reply:
The `stateLock` is protecting the transactionMetadataCache, for put / get access; within the cache any in-place updates of the metadata object need to be protected by a the metadata object lock itself. The introduction of the `TransitionMetadata` is to avoid in-place updates during the prepare update period.

Let me know if you find any corner cases that will introduce a race condition.

Reply:
Where is the `stateLock` acquired? Between updating the `txnMetadata` in place and actually writing the update to the log, can another request modify the metadata? If not, what locking prevents it? Or is it impossible for another request to come in at this time? 

I don't think we need to block this PR on these questions, since the pattern is everywhere. I am probably missing something in the broader locking context.